### PR TITLE
[feat] refactor the settings page

### PR DESCRIPTION
**Pull request by jinkang-0**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

tl;dr: refactored settings.

New components:
- `@/components/SettingsSection`
    - `SettingSection` - big card that wraps each section in settings
    - `SettingField` - each field in the section (i.e. first name)
    - sort of extends forms to provide markup templating to build each section (more readable than JSON object)
- `@/components/EditButton`
    - reusable edit button for settings and onboarding review
    - two types: link, regular button
- `@/components/Buttons`
    - added red button and regular async button (not beeg)

Refactors:
- Refactored Zod schemas into `@/data/formSchemas.ts`
- Added `ProfileToUpload` schema type that extends the regular `Profile` but also accepts `null` values for optional values (used to upsert `NULL` into Supabase) and use `Date` instead of `string` for timestamp fields

Minor tweaks:
- Reconfigured redirects:
    - Settings page, not logged in -> login page
    - Onboarding, not logged in -> login page (instead of home page)
    - Onboarding, already onboarded -> settings page (instead of home page)
- Added hook `useProfileAuth` to get both `auth` and `profile` contexts, whilst checking:
    1. that they exist (i.e. the component using the hook must be wrapped in the context provider)
    2. that the user is logged in (if not, redirect to login)

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @kyrenetam at the bottom."

image

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

<details>
<summary>Git Prereqs</summary>

```bash
git fetch origin jinkang/ijp-127-refactor-the-settings-page
git checkout jinkang/ijp-127-refactor-the-settings-page
git pull origin jinkang/ijp-127-refactor-the-settings-page
npm run dev
```
</details>

Recommended order:
- `@/components/SettingsSection/index.tsx`
- `@/app/settings/page.tsx`
- `@/components/SettingsSection/...`
    - `.../BasicInformationSection.tsx`
    - `.../AvailabilitySection.tsx`
    - `.../RolesSection.tsx`

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @varortz, @kyrenetam

### CODE REVIEW

File: src/api/supabase/queries/profiles.ts [hunk 1]
```diff
@@ -130,7 +135,10 @@ export async function deleteRoles(userId: UUID, deleteInfo: ProfileRole[]) {
     .from('profiles-roles')
     .delete()
     .eq('user_id', userId)
-    .in('role', Array.from(new Set(deleteInfo.map(i => i.role))));
+    .in(
```
Comment (ccatherinetan) [inline src/api/supabase/queries/profiles.ts:138]: What does .in() do again?

File: src/app/onboarding/availability/page.tsx [hunk 2]
```diff
@@ -11,34 +11,22 @@ import { FormControl, FormField, FormItem, FormLabel } from '@/components/Form';
...
+  parseDateString,
+} from '@/utils/helpers';
 import {
   useGuardedOnboarding,
   useOnboardingNavigation,
   useScrollToTop,
 } from '@/utils/hooks';
 import * as Styles from '../styles';
 
-// define form schema to automate form validation
```
Comment (ccatherinetan) [inline src/app/onboarding/availability/page.tsx:24]: why was this removed?

File: src/app/onboarding/legal-credentials/page.tsx [hunk 3]
```diff
@@ -51,12 +38,12 @@ export default function Page() {
...
   // initialize form with data from onboarding context
-  const form = useForm<z.infer<typeof legalExperienceSchema>>({
-    resolver: zodResolver(legalExperienceSchema),
+  const form = useForm<z.infer<typeof legalCredentialSchema>>({
+    resolver: zodResolver(legalCredentialSchema),
     defaultValues: {
-      stateBarred: onboarding.profile.state_barred,
-      barNumber: onboarding.profile.bar_number,
-      eoirRegistered: onboarding.profile.eoir_registered,
+      stateBarred: onboarding.profile.state_barred ?? undefined,
```
Comment (ccatherinetan) [inline src/app/onboarding/legal-credentials/page.tsx:44]: Did we do this to ensure that it's not null?

File: src/app/onboarding/legal-fellow-credentials/page.tsx [hunk 4]
```diff
@@ -42,17 +36,19 @@ export default function Page() {
...
 
   // initialize form with values from onboarding context
-  const form = useForm<z.infer<typeof legalExperienceSchema>>({
-    resolver: zodResolver(legalExperienceSchema),
+  const form = useForm<z.infer<typeof legalFellowCredentialSchema>>({
+    resolver: zodResolver(legalFellowCredentialSchema),
     defaultValues: {
       expectedBarDate: onboarding.profile.expected_bar_date
-        ? new Date(`${onboarding.profile.expected_bar_date}T00:00`)
+        ? onboarding.profile.expected_bar_date
```
Comment (ccatherinetan) [inline src/app/onboarding/legal-fellow-credentials/page.tsx:49]: was there a schema change that warranted this change?

File: src/components/SettingsSection/RolesSection.tsx [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+  const onValidSubmit = async (values: z.infer<typeof roleAndLegalSchema>) => {
+    const { userId } = auth;
+    if (!userId) throw new Error('User ID not found! Are you logged in?');
+
+    // parse roles to update into array
+    // nasty boolean checking to ensure type safety
+    const rolesToUpdate: ProfileRole[] = [];
+
+    const { isAttorney, isLegalFellow, isInterpreter } = getRoleBools(
```
Comment (varortz) [inline src/components/SettingsSection/RolesSection.tsx]: is this helper necessary here? This seems to also be the only be the only place the helper is used.
Can't we just do the check in the lines below? 

`if (values.roles.includes('ATTORNEY'))`, etc.

File: src/data/formSchemas.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+  startDate: z
+    .date({
+      required_error:
+        'Please include your estimated starting date of availability',
+    })
+    .min(getCurrentDate(), { message: 'Must select a current or future date' }),
+  availability: z.string().optional().nullable(),
+});
+
+export const legalCredentialSchema = z.object({
```
Comment (varortz) [inline src/data/formSchemas.ts]: nit: can we update this to `attorneyCredentialSchema` to make it consistent with `legalFellowCredentialSchema`? the naming can be a bit confusing otherwise

File: src/api/supabase/queries/profiles.ts (hunk 1 above)
Comment (jinkang-0) [reply src/api/supabase/queries/profiles.ts:138]: it's a filter like `.eq`, basically filters only rows whose value for `role` is within the specified array
for more information, see [official documentations](https://supabase.com/docs/reference/javascript/in)

File: src/app/onboarding/legal-credentials/page.tsx (hunk 3 above)
Comment (jinkang-0) [reply src/app/onboarding/legal-credentials/page.tsx:44]: yer

File: src/app/onboarding/legal-fellow-credentials/page.tsx (hunk 4 above)
Comment (jinkang-0) [reply src/app/onboarding/legal-fellow-credentials/page.tsx:49]: ah yes, recall the PR description:
> Added ProfileToUpload schema type that extends the regular Profile but also accepts null values for optional values (used to upsert NULL into Supabase)

I forgot to add that this also included changing timestamp `string` into just JS `Date` objects, which was more compatible with `timestamptz` type on Supabase.

File: src/components/SettingsSection/RolesSection.tsx (hunk 5 above)
Comment (jinkang-0) [reply src/components/SettingsSection/RolesSection.tsx]: yeah lowkey thought id need it in more places
i can remove it lol

**Review by varortz — COMMENTED** [review]
WE'RE FINALLY DONE WITH REFACTORING NOW!! Thank you so much jinkang, looks really good, just left a couple comments. Once they're addressed, I can approve.

**Review by varortz — APPROVED** [review]
lgtm!

File: src/components/SettingsSection/BasicInformationSection.tsx [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    values.canReads.forEach(lang => {
+      try {
+        allLangs[lang.label].read = true;
+      } catch {
+        allLangs[lang.label] = { read: true, speak: false };
+      }
+    });
+    values.canSpeaks.forEach(lang => {
+      try {
+        allLangs[lang.label].speak = true;
```
Comment (ccatherinetan) [inline src/components/SettingsSection/BasicInformationSection.tsx:78]: can i ask when/why this assignment would error?

**Review by ccatherinetan — APPROVED** [review]
looks good!

File: src/components/SettingsSection/BasicInformationSection.tsx (hunk 7 above)
Comment (jinkang-0) [reply src/components/SettingsSection/BasicInformationSection.tsx:78]: ye this can error if `allLangs[lang.label]` is undefined (i.e. the key references nothing), so attempting to do `undefined.speak` would error
